README.md
README
================

  - [Power2Stage](#power2stage)
      - [Supported Methods](#supported-methods)
          - [Simulation-based](#simulation-based)
              - [‘Type 1’](#type-1)
              - [‘Type 2’](#type-2)
              - [Sample Size Re-estimation in the
                Interim](#sample-size-re-estimation-in-the-interim)
              - [Group Sequential Design](#group-sequential-design)
          - [Inverse-Normal Combination](#inverse-normal-combination)
      - [Functions](#functions)
          - [Main](#main)
          - [Helpers](#helpers)
      - [Examples](#examples)
          - [Method B](#method-b)
          - [Method C](#method-c)
          - [Inverse-Normal Combination](#inverse-normal-combination-1)
      - [Speed Comparisons](#speed-comparisons)
      - [Installation](#installation)

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![cran
checks](https://cranchecks.info/badges/summary/PowerTOST)](https://cran.r-project.org/web/checks/check_results_Power2Stage.html)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/Power2Stage?color=blue)](https://r-pkg.org/pkg/Power2Stage)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/last-month/Power2Stage?color=green)](https://r-pkg.org/pkg/Power2Stage)

# Power2Stage

The package contains functions to obtain the operational characteristics
(power, type I error, percentage of studies proceeding to the second
stage, average and quantiles of total sample sizes) of bioequivalence
studies in adaptive sequential 2-Stage Designs (TSD) via simulations.  
Built 2019-09-15 with R 3.6.1.

## Supported Methods

### Simulation-based

Since the many letters denoting the methods given by various authors
might be confusing, [I
classified](https://doi.org/10.1007/s00228-015-1806-2) the methods as
two ‘types’:

  - **‘Type 1’**  
    An adjusted *α* is used *both* in the interim as well as in the
    final analysis of pooled data.
  - **‘Type 2’**  
    Whether an unadjusted or an adjusted *α* is used depends on interim
    power. An adjusted *α* is used in the final analysis of pooled data.

It should be noted that the adjusted alphas do not necessarily have to
be the same in both stages. Below a summary of conditions used in the
decision schemes of the published methods.

<small>[TOC ↩](#readme)</small>

#### ‘Type 1’

  - [Potvin *et al.*](https://doi.org/10.1002/pst.294) (2008) ‘Method
    B’: *α* 0.0294 (*θ*<sub>0</sub> 0.95, target power 0.80).
  - [Fuglsang](https://doi.org/10.1208/s12248-013-9475-5) (2013) ‘Method
    B’: *α* 0.0284 (*θ*<sub>0</sub> 0.95, target power 0.90).
  - [Karalis](https://doi.org/10.1016/j.ijpharm.2013.08.013) (2013)
    ‘TSD-2’: *α* 0.0294 (*θ*<sub>0</sub> = PE, target power 0.80).
  - [Fuglsang](https://doi.org/10.1208/s12248-014-9571-1) (2014) ‘Method
    B’ (parallel design): *α* 0.0294 (*θ*<sub>0</sub> 0.95, target power
    0.80).
  - [Zheng *et al.*](https://doi.org/10.1002/pst.1672) (2015) ‘MSDBE’:
    *α*<sub>1</sub> 0.01, *α*<sub>2</sub> 0.04.
  - [Xu *et al.*](https://doi.org/10.1002/pst.1721) (2016) ‘Method E’:
    (*θ*<sub>0</sub> 0.95, target power 0.80, *n<sub>max</sub>* 42).  
      - For *CV* 10–30%  
        *α*<sub>1</sub> 0.0294, *α*<sub>2</sub> 0.0357, futility rule on
        CI {0.9374, 1/0.9374}.
      - For *CV* 30–55%  
        *α*<sub>1</sub> 0.0254, *α*<sub>2</sub> 0.0363, futility rule on
        CI {0.9305, 1/0.9305}.
  - [Molins *et al.*](https://doi.org/10.1002/sim.7452) (2017) ‘Type 1
    modified Potvin B’: *α* 0.0301 (*θ*<sub>0</sub> 0.95, target power
    0.80, min. *n<sub>2</sub>* = 1.5*n<sub>1</sub>*, *n<sub>max</sub>*
    150).

#### ‘Type 2’

  - [Potvin *et al.*](https://doi.org/10.1002/pst.294) (2008) ‘Method
    C’: *α* 0.0294 (*θ*<sub>0</sub> 0.95, target power 0.80).
  - [Montague *et al.*](https://doi.org/10.1002/pst.483) (2011) ‘Method
    D’: *α* 0.0280 (*θ*<sub>0</sub> 0.90, target power 0.80).
  - [Fuglsang](https://doi.org/10.1208/s12248-013-9475-5) (2013) ‘Method
    C/D’:  
    *α* 0.0274 (*θ*<sub>0</sub> 0.95, target power 0.90).  
    *α* 0.0269 (*θ*<sub>0</sub> 0.90, target power 0.90).
  - [Karalis and Macheras](https://doi.org/10.1007/s11095-013-1026-3)
    (2013) ‘TSD’: *α* 0.0294 (*θ*<sub>0</sub> = PE, target power 0.80).
  - [Karalis](https://doi.org/10.1016/j.ijpharm.2013.08.013) (2013)
    ‘TSD-1’: *α* 0.0280 (*θ*<sub>0</sub> = PE, target power 0.80).
  - [Xu *et al.*](https://doi.org/10.1002/pst.1721) (2016) ‘Method F’:
    (*θ*<sub>0</sub> 0.95, target power 0.80, *n<sub>max</sub>* 180).  
      - For *CV* 10–30%  
        *α*<sub>1</sub> 0.0248, *α*<sub>2</sub> 0.0364, futility rule on
        CI {0.9492, 1/0.9492}.
      - For *CV* 30–55%  
        *α*<sub>1</sub> 0.0259, *α*<sub>2</sub> 0.0349, futility rule on
        CI {0.9350, 1/0.9350}.
  - [Molins *et al.*](https://doi.org/10.1002/sim.7452) (2017) ‘Type 2
    modified Potvin C’: *α* 0.0280 (*θ*<sub>0</sub> 0.95, target power
    0.80, min. *n<sub>2</sub>* = 1.5*n<sub>1</sub>*, *n<sub>max</sub>*
    150).

#### Sample Size Re-estimation in the Interim

[Golkowski *et al.*](https://doi.org/10.1002/pst.1617) (2014).

#### Group Sequential Design

[Kieser and Rauch](https://doi.org/10.1002/sim.6487) (2015).

### Inverse-Normal Combination

[König *et al.*](https://doi.org/10.13140/RG.2.1.5190.0967) (2014),
[Wassmer and Brannath](https://doi.org/10.1007/978-3-319-32562-0)
(2016), [Maurer *et al.*](https://doi.org/10.1002/sim.7614) (2018).

<small>[TOC ↩](#readme)</small>

## Functions

### Main

Defaults employed if not specified in the function call:

| function          | `theta0` | `target power` | `usePE` | `Nmax` | `max.n` | `fCrit` | `fClower` |
| ----------------- | :------: | :------------: | :-----: | :----: | :-----: | :-----: | :-------: |
| `power.tsd()`     |   0.95   |      0.80      | `FALSE` |  Inf   |    –    |    –    |     –     |
| `power.tsd.fC()`  |   0.95   |      0.80      | `FALSE` |   –    |   Inf   | `"PE"`  |   0.80    |
| `power.tsd.KM()`  |   0.95   |      0.80      |    –    |  150   |    –    |    –    |     –     |
| `power.tsd.ssr()` |   0.95   |      0.80      | `FALSE` |   –    |   Inf   |    –    |     –     |
| `power.tsd.GS()`  |   0.95   |       –        |    –    |   –    |    –    | `"PE"`  |   0.80    |
| `power.tsd.in()`  |   0.95   |      0.80      | `FALSE` |   –    |   Inf   | `"CI"`  |   0.95    |
| `power.tsd.p()`   |   0.95   |      0.80      | `FALSE` |  Inf   |    –    |    –    |     –     |

All functions are for a 2×2×2 crossover design except `power.tsd.p()`,
which is for a two-group parallel design.  
In all methods `alpha=c(0.0294, 0.0294)`. In ‘Type I’ designs only
(`method="C"`) additionally `alpha0=0.05`.  
If `usePE=TRUE` the point estimate in the interim is used in sample size
estimation of the second stage.  
If the estimated total sample size exceeds `max.n` the second stage will
be forced to `max.n-n1` (*i.e.*, it is *not* a futility criterion).  
The method used for interim power and sample size estimation is
specified by the argument `pmethod`. It defaults to `"nct"`
(approximation by the noncentral *t*-distribution) except in
`power.tsd.GS()`, where the total sample size is already fixed.  
The BE limits are specified by the arguments `theta1` and `theta2`
(default to 0.80 and 1.25).  
The number of simulations is specified by the argument `nsims`. It
defaults to 1e5 if simulating power and to 1e6 if simulating the empiric
type I error (*i.e.*, `theta0` set to the value of `theta1` or
`theta2`).

**Futility Criteria in the Interim**

  - `Nmax`: The study will stop if the estimated total sample size
    exceeds `Nmax`.
  - `fCrit` (`"PE"` or `"CI"`): The study will stop if outside `fClower`
    and `1/fClower`.
      - `"PE"`: `fClower` defaults to 0.80.
      - `"CI"`: `fClower` defaults to 0.925 (except in function
        `power.tsd.in()`, where it defaults to 0.95).

<small>[TOC ↩](#readme)</small>

### Helpers

  - `sampleN2.TOST()`  
    Estimates the sample size of stage 2 to achieve at least the target
    power.
  - `interim.tsd.in()`  
    Interim analysis based on the Inverse-Normal Combination method.
  - `final.tsd.in()`  
    Final analysis based on the Inverse-Normal Combination method.

<small>[TOC ↩](#readme)</small>

## Examples

Before running the examples attach the library.

``` r
library(Power2Stage)
```

If not noted otherwise, defaults are employed.

### Method B

Power estimation by the shifted central *t*-distribution.

``` r
power.tsd(CV = 0.20, n1 = 12, pmethod = "shifted")
# TSD with 2x2 crossover 
# Method B: alpha (s1/s2) = 0.0294 0.0294 
# Target power in power monitoring and sample size est. = 0.8
# Power calculation via shifted central t approx. 
# CV1 and GMR = 0.95 in sample size est. used
# No futility criterion
# BE acceptance range = 0.8 ... 1.25
# 
# CV = 0.2; n(stage 1) = 12; GMR = 0.95
# 
# 1e+05 sims at theta0 = 0.95 (p(BE) = 'power').
# p(BE)    = 0.84454
# p(BE) s1 = 0.41333
# Studies in stage 2 = 56.45%
# 
# Distribution of n(total)
# - mean (range) = 20.7 (12 ... 82)
# - percentiles
#  5% 50% 95% 
#  12  18  40
```

Explore the empiric type I error at the upper BE-limit.

``` r
power.tsd(CV = 0.20, n1 = 12, pmethod = "shifted",
          theta0 = 1.25)[["pBE"]]
# [1] 0.046352
```

<small>[TOC ↩](#readme)</small>

### Method C

Power estimation by the shifted central *t*-distribution.

``` r
power.tsd(method = "C", CV = 0.20, n1 = 12, pmethod = "shifted")
# TSD with 2x2 crossover 
# Method C: alpha0 = 0.05, alpha (s1/s2) = 0.0294 0.0294 
# Target power in power monitoring and sample size est. = 0.8
# Power calculation via shifted central t approx. 
# CV1 and GMR = 0.95 in sample size est. used
# No futility criterion
# BE acceptance range = 0.8 ... 1.25
# 
# CV = 0.2; n(stage 1) = 12; GMR = 0.95
# 
# 1e+05 sims at theta0 = 0.95 (p(BE) = 'power').
# p(BE)    = 0.8496
# p(BE) s1 = 0.42656
# Studies in stage 2 = 53.7%
# 
# Distribution of n(total)
# - mean (range) = 20.6 (12 ... 82)
# - percentiles
#  5% 50% 95% 
#  12  18  40
```

Slightly better than ‘Method B’ in terms of power in both stages and
fewer studies are expected to proceed to the second stage.

Explore the empiric type I error at the upper BE-limit (1 milion
simulations).

``` r
power.tsd(method = "C", CV = 0.20, n1 = 12, pmethod = "shifted",
          theta0 = 1.25)[["pBE"]]
# [1] 0.051238
```

Slight inflation of the type I error (although considered negligible by
the authors). However, more adjustment (adjusted *α* 0.0280) controls
the type I error.

``` r
power.tsd(method = "C", alpha = rep(0.0280, 2), CV = 0.20,
          n1 = 12, pmethod = "shifted", theta0 = 1.25)[["pBE"]]
# [1] 0.049903
```

<small>[TOC ↩](#readme)</small>

### Inverse-Normal Combination

Data given by Potvin *et al.* in Example 2: 12 subjects in stage 1, PE
1.0876, CV 0.18213, all defaults of the function used.

``` r
interim.tsd.in(GMR1 = 1.0876, CV1 = 0.18213, n1= 12)
# TSD with 2x2 crossover
# Inverse Normal approach
#  - Maximum combination test with weights for stage 1 = 0.5 0.25 
#  - Significance levels (s1/s2) = 0.02635 0.02635 
#  - Critical values (s1/s2) = 1.93741 1.93741 
#  - BE acceptance range = 0.8 ... 1.25
#  - Observed point estimate from stage 1 is not used for SSR
#  - With conditional error rates and conditional estimated target power
# 
# Interim analysis after first stage
# - Derived key statistics:
#   z1 = 3.10000, z2 = 1.70344,
#   Repeated CI = (0.92491, 1.27891)
# - No futility criterion met
# - Test for BE not positive (not considering any futility rule)
# - Calculated n2 = 6
# - Decision: Continue to stage 2 with 6 subjects
```

The second stage should be initiated with 6 subjects. Note that with
`interim.tsd.in(..., fCrit="No", ssr.conditional="no")` 8 subjects would
be required like in the Methods of Potvin *et al.*

The second stage is performed in 8 subjects, PE 0.9141, CV 0.25618.

``` r
final.tsd.in(GMR1 = 1.0876, CV1 = 0.18213, n1 = 12,
             GMR2 = 0.9141, CV2 = 0.25618, n2 = 8)
# TSD with 2x2 crossover
# Inverse Normal approach
#  - Maximum combination test with weights for stage 1 = 0.5 0.25 
#  - Significance levels (s1/s2) = 0.02635 0.02635 
#  - Critical values (s1/s2) = 1.93741 1.93741 
#  - BE acceptance range = 0.8 ... 1.25
# 
# Final analysis after second stage
# - Derived key statistics:
#   z1 = 2.87952, z2 = 2.60501,
#   Repeated CI = (0.87690, 1.17356)
#   Median unbiased estimate = 1.0135
# - Decision: BE achieved
```

The study passed with a 100(1 – 2×0.02635) = 94.73% CI of 87.69–117.36%.
Although slightly more conservative, same conclusion like based on the
94.12% CI of 88.45–116.38% reported by Potvin *et al.*

<small>[TOC ↩](#readme)</small>

## Speed Comparisons

Performed on a double Xeon E3-1245v3 3.4 GHz, 8 MB cache, 16 GB RAM, R
3.6.1 64 bit on Windows 7.

‘Method B’ (*CV* 0.20, *n*<sub>1</sub> 12).

    #   method   power seconds
    #  shifted 0.84454    1.09
    #      nct 0.84266    1.61
    #    exact 0.84260   31.98

Despite being the fastest, the shifted central *t*-distribution should
only be used in order to compare with published methods. The noncentral
*t*-distribution is a good compromise between speed and accuracy and
hence, the default in all functions. The exact method based on Owen’s
Q-function is time-consuming and therefore, not recommended in
validating a custom method in a narrow grid of
*n*<sub>1</sub>/*CV*-combinations. However, in designing a new study it
is the method of choice.

Blinded sample size re-estimation (*α* 0.03505, *CV* 0.239,
*n*<sub>1</sub> 10, target power 0.90), 1 million simulations for the
empiric type I error.

    #   method      TIE seconds
    #       ls 0.049054    3.67
    #  shifted 0.046106   12.85
    #      nct 0.046319   18.24
    #    exact 0.046319  429.10

The crude large sample approximation (`pmethod="ls"`) should only be
used to compare with the published method.

<small>[TOC ↩](#readme)</small>

## Installation

You can install the released version of Power2Stage from
[CRAN](https://CRAN.R-project.org) with:

``` r
package <- "Power2Stage"
inst    <- package %in% installed.packages()
if (length(package[!inst]) > 0) install.packages(package[!inst])
```

And the development version from [GitHub](https://github.com/) with:

    # install.packages("devtools")
    devtools::install_github("Detlew/Power2Stage")

<small>[TOC ↩](#readme)</small>


NEWS
version 0.5.2.9000

Version 0.5.2 (easter egg)
2019-04-21
   - Maintenance release with mainly bug fixes.
   - Bug fix in calculation of median unbiased estimate within final.tsd.in().
   - Typos and minor wording and formatting corrections (man pages and x.tds.in()).
   - Bug fix in power.tsd.GS (regarding variable hw).
   - Bug in interim.tsd.in() w.r.t to power > target triggering
     futulity fixed (THX to mittyri).

Version 0.5-1 (belated easter egg)
2018-04-03
   - All functions xyz.2stage.ab() now have an alias xyz.tsd.ab(), e.g. instead
     of power.2stage.fC() one may use power.tsd.fC().
     The versions with .2stage. in their names will be removed in later versions.
   - The futility rule for a CI in power.2stage.GS() now based on 90% CI
     to be in line with the other functions using such a futility rule.
   - Default fCrit in power.2stage.GS() changed to "CI".
   - Added interim.2stage.in() and final.2stage.in() to perform interim and
     final analysis of 2 stage designs based on the Standard Combination or
     Maximum Combination Test (also includes a print class).
   - Update defaults of power.2stage.in() in order to be consistent with
     Paper from Maurer et al (2018).
   - Aliases power.tsd.xyz introduced (not public yet).
   - Experimental function power.tsd.2m() to deal with 2 PK metrics
     in the TSD decision scheme Potvin "B". Not public yet.
Version 0.4-6
2017-10-27
   - Updates of various man pages.
   - New function sampleN2.TOST() to estimate n2 in 2x2 and parallel designs
     based on df=n-3 for stage 2.
   - df for sample size re-estimation changed to n-3 for Potvin like TSD schemes
     (pooled evaluation with additionally stage term)
   - Same for power.2Stage.ssr().
   - Function power.2stage.in() to calculate the operational characteristics of
     TSDs evaluated via p-value combination using the inverse normal approach.
     Contributed by Benjamin Lang.
   - GitHub URL and bug reports URL added.
Version 0.4-5
2017-01-18
   - DOI for references in man pages added.
   - Bug in power.2stage.p() and power.2stage.pAF() removed which causes
     negative residual variances in case of test="anova" and imprecise power
     values in all other cases for certain scenarios of n1, CV.
   - Large sample approximation to the sample size adapted to deal with
     unsymmetric BE acceptance ranges, one limit = Inf allowed
   - Maintenance release to refelect changes in power calculation of
     package PowerTOST
   - Some small cosmetic changes in code w.r.t min.n2
Version 0.4-4
2016-01-21 (not released via CRAN)
   - Method "B0" (re)introduced in power.2stage() and power.2stage.fC() which
     employs the decision scheme of the so-called MSDBE (Modified Sequential
     Design for BE) of Zheng et al.
   - Method "B" in power.2stage.p() / power.2stage.pAF() also adapted to
     decision scheme E of Xu et al.
   - Figures with TSD decision schemes added in /doc subdirectory.
Version 0.4-3
2015-11-24
   - power2stage.fC() now has a new argument max.n which constraints the total
     sample size
   - Method "B" in power2stage.fC() and power.2stage() is redefined to deal
     with unequal alpha[1] and alpha[2] in the same manner as method "E" in
     Xu et al.This gives small differences compared to previous calculations
     with unequal alphas, f.i. Haybittle-Peto alphas
   - The CI futility criterion is now based on the 90% CI according to Xu et al.
     Was formerly the 1-2*alpha[1] CI
   - Default futility criterion changed to "CI" with a lower bound = 0.925
   - Defunct print argument removed from all power.2stage.xyz() functions
Version 0.4-2
2015-07-13
   - Bug removed in power.2stage.ssr() which prevented correct sample-size
     re-estimation if power was calculated via pmethod="ls"
   - power.2stage.ssr() now has an argument usePE to use the point estimate from
     the interim analysis in sample-size re-estimation (makes only sense if
     blind=FALSE)
Version 0.4-1
2015-06-19
   - Major rewrite of the power.2stage.xyz() functions wich now return an S3 object
     of class "pwrtsd". Output is now done more R-like via the S3 print method for
     class "pwrtsd". Therefore the print argument is defunc and will be removed
     in the next version
     The detail argument has now default=FALSE. Thus be patient if you simulate
     for alpha with 1 Mio sims
   - The power.2stage.xyz() functions (with exception of power.2stage.GS()) now
     return a component ntable, a 'table' object for summarizing the discrete
     distribution of ntotal. The nhist component was removed
   - The default of fClower in case of fCrit="PE" was changed to 0.8 in function
     power.2stage.fC()
   - The nsims argument, if missing, is now set to 1E6 if you simulate for alpha
     (i.e. with theta0 at border or outside acceptance range theta1 ... theta2)
     and to 1E5 otherwise
Version 0.3-2
2015-06-10 released to alpha tester(s) only
   - The power.2stage.yy() functions (with exception of power.2stage.GS() now
     return a component nhist with class "histogram" which can be used with
     plot() to visualize the distribution of Ntotal (suggested by H. Schuetz)
   - Slight improvements and typo fixes in man pages
   - Handling of non-integer df's in power.2stage.p() in case of Welch's test
     changed (now no truncation)
   - power.2stage.fC() now has also power calculation method "shifted" implemented
Version 0.3-1
2015-01-24  (dedicated to my brothers Stefan's 50 birthday)
   - Major code revision to reflect the internal changes in upcoming
     PowerTOST V1.2-06 'raw' power functions
   - Sample-size estimation routine (vectorized!) reworked which gives
     a considerable run-time boost
   - Internal change of start value of sample size estimation to Zhang's
     formula (may change the extremal value of n and its mean if usePE=TRUE)
   - Function power.2stage.p() accepts now unbalanced stage 1 and uses the

R/SampleN2.TOST.R
# ----------------------------------------------------------------
# function to estimate the sample size of the second stage (in the
# pooled analysis one degree of freedom less than in the common
# model due to the 'stage'-term).
#
# Author: H. Schuetz, modified by D. Labes
# ----------------------------------------------------------------
sampleN2.TOST <- function(alpha=0.0294, CV, n1, theta0=0.95,
                          theta1=0.8, theta2=1.25, targetpower=0.8,
                          design="2x2", method="exact", imax=100)
  {
  if (missing(CV) || CV <=0) stop("CV must be given and positive.")
  if (missing(n1) || n1 < 4) stop("n1 must be given and >=4.")
  if (targetpower <=0 || targetpower >= 1) stop("targetpower must be >0 and <1.")
  design <- tolower(design)
  if (!design %in% c("2x2", "2x2x2", "parallel"))
     stop("design must be any of \"2x2\", \"2x2x2\", or \"parallel\"")
  if(theta0 < theta1 || theta0 > theta2) stop("theta0 not within BE limits.")
  # design constant in terms of ntot
  ifelse (design == "parallel", bk <- 4, bk <- 2)
  dfe   <- expression(n - 3) # pooled evaluation with additional 'stage' effect
  steps <- 2 # stepsize for sample size search
  nmin  <- 4 # minimum n
  ltheta1 <- log(theta1)
  ltheta2 <- log(theta2)
  diffm   <- log(theta0)
  se      <- CV2se(CV)
  # start value by Zhang's method
  n       <- .sampleN0_3(alpha, targetpower, ltheta1, ltheta2, diffm, se, steps, bk)
  if (n < nmin) n <- nmin
  df  <- eval(dfe)
  pow <- .calc.power(alpha, ltheta1, ltheta2, diffm, sem=se*sqrt(bk/n), df, method)
  iter <- 0
  down <- FALSE; up <- FALSE
  while (pow > targetpower) {
    if (n <= nmin) break
    down <- TRUE
    n    <- n - steps
    iter <- iter + 1
    df   <- eval(dfe)
    pow  <- .calc.power(alpha, ltheta1, ltheta2, diffm, sem=se*sqrt(bk/n), df, method)
    if (iter > imax) break
  }
  while (pow < targetpower) {
    up   <- TRUE; down <- FALSE
    n    <- n + steps
    iter <- iter + 1
    df   <- eval(dfe)
    pow  <- .calc.power(alpha, ltheta1, ltheta2, diffm, sem=se*sqrt(bk/n), df, method)
    if (iter > imax) break
  }
  # total sample size to stage 2 sample size
  ifelse (n - n1 > 0, n2 <- n - n1, n2 <- 0)
  if (n2 > 0) {
    pow <- .calc.power(alpha, ltheta1, ltheta2, diffm, sem=se*sqrt(bk/(n1+n2)),
                       df=eval(n1+n2), method)
  } else {
    # usual df if study stops in stage 1
    pow <- power.TOST(alpha=alpha, theta1=theta1, theta2=theta2,
                      theta0=theta0, CV=CV, n=n1, method=method)
  }
  res <- data.frame(Design=design, alpha=alpha, CV=CV, theta0=theta0,
                    theta1=theta1, theta2=theta2, n1=n1, n2=n2, pow=pow,
                    target=targetpower)
  names(res) <- c("Design", "alpha", "CV", "theta0", "theta1", "theta2",
                  "n1", "Sample size", "Achieved power", "Target power")
  return(res)
}
